Expose node token and range information (JAVA-312).

olim7t opened the pull request:
New features:
- `Token` class now public, new `TokenRange` class
- `Host#getTokens()`
- new methods in `Metadata`: `getTokenRanges()`, `getTokenRanges(String, Host)`, `getReplicas(String, TokenRange)`.
- using `Token` as a "pseudo data-type" in queries: `Row#getToken`, `BoundStatement#setToken`, using a token as an argument in a `SimpleStatement`.

olim7t committed:
Expose node token and range information (JAVA-312).

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/ArrayBackedRow.java:
```diff
@@ -239,6 +242,23 @@ public InetAddress getInet(String name) {
+    }
+
+    public Token getToken(String name) {
+        return getToken(metadata.getFirstIdx(name));
```
I've noticed that for this to work, you need to pass in the full name of the selected value.  I.e. if doing:

```
select token(i) from test.foo where token i = 1
```

you would need to call `row.getToken("token(i)");`, is that the intent of this?  Should we change getToken(name) to implicitly add 'token(" + name + ")" or is it fine the way it is?  I could see it be misleading either way, so adding a note to the javadoc would be good.

EDIT: Thinking about it more, I think I like this the way it is since it is consistent with the rest of the methods.

olim7t replied:
The getter retrieves a column of a type that matches the token type for the current partitioner. The fact that it was produced with the `token()` function is a detail that should not have an impact (even though it will be the most common case and I can't think of any other meaningful use at the moment).

pcmanus replied:
Note sure I understand that last reasoning tbh, but I tend to agree with @tolbertam here: 99.999% of the time, you'll call the `getToken` getter for a column on which you had applied the `token` function in the query. And really, that's the only useful case. That is, in theory you coul call `getToken` for any column as long as the type happens to match the one of the tokens for this partitioner, but why would you ever need that?

So I do think we should automatically add the `token` part inside the getToken call. Having to always write `getToken("token(i)")` is redundant and a bit ridiculous. Note that I'm fine making it smarter so that it work whether or not you give the "token()" part yourself or not, but imo it's not worth the complexity.

olim7t replied:
OK, as I said I agree that there's probably no valid case for calling it on a regular column.

I'll make the change, and we'll add the smarter logic if someone asks for it.

olim7t replied:
Actually, there is a legitimate case, when you alias the column:

```
select token(i) AS my_token from test.foo where token i = 1
```

So I guess I'll do the "smart" version directly.

olim7t replied:
Mmm, after all I'm not sure this is a good idea, because we must also consider that case:

```
select k, token(k) from foo where k = 1;
```

pcmanus replied:
I don't really see the problem: what `getToken(String name)` should do is first check if it can find `"token(" + name + ")"`, and if it can't, it checks name directly for the case where we've used an alias. If we have both "k" and "token(k)" in the metadata and you call `getToken("k")`, there is virtually no change you meant to fetch and try to deserialze "k" as a token, especially since in most case it won't work anyway.

Additionally, we could first check if `name.startsWith("token(")` and avoid trying to prepend in that case, which is what I called the "smarter" version, but that part if optional imo.

tolbertam replied:
Should have thought to mention this, should this same logic be applied to BoundStatement#setToken(name, token)?

olim7t replied:
When the query string contains `token(i) = ?`, the variable name in the prepared statement's metadata is "partition key token", so that wouldn't work.
I think set by name is more likely to be used when the query string uses named placeholders, like in `where token(i) = :myvalue`.

pcmanus replied:
Cassandra isn't terribly consistent on that front, but I do think nonetheless that we must make sure `getToken` and `setToken` are as symetric and easy to use in practice as possible

That being said, we do have a problem even with {{getToken}} and it's that it won't really work for a compound partition key, since the token apply to all partition key columns, not just one. Not sure how we can fix that, but I do strongly think that having to write `getToken("token(i)")` to get `i`'s token, and `setToken("partition key token")` to set it is going to confuse the hell out of people.

One thing to take into account however is that in practice, the only useful use of the `token()` method is applied to all partition keys. And so imo it's ok to only support that case out of the box and ignore weird use of token if that help make things simpler. For instance, maybe we could have a magic constant `PARTITION_KEY` that both `getToken` and `setToken` would recognize and would do the right thing with. I'm not 100% satisfied by the idea though so open to better ideas.

olim7t replied:
I did eventually implement `getToken("i")` per your suggestion, sorry to not have said it explicitly in a comment.
Regarding `setToken`, my impression is that if you use `?` placeholders you tend to use index-based setters, so the weird column name wouldn't be that big of a problem.

pcmanus replied:
> I did eventually implement getToken("i") per your suggestion

Yes, but what I'm saying is that my suggestion is actually broken for compound partition keys, which is really not ideal.

> my impression is that if you use ? placeholders you tend to use index-based setters, so the weird column name wouldn't be that big of a problem

I'm not sure where that impression comes from but I strongly suspect its wrong, if only because there is plenty of example on the web that use `?` placeholders and name-based setters. And for normal columns that combination works pretty well. In fact, I rarely see people using index-based setters (because it's not terribly friendly), and I don't see everyone using named bind markers. Anyway, I don't have numbers to back any of this but I have a strong hunch that you are overly optimistic on that front.

Anyway, I do worry that having non-symetric `getToken` and `setToken` will seriously surprise people and break expectations.

olim7t replied:
As an alternative to the magic constant, we could have dedicated methods (`getPartitionKeyToken()` and `setPartitionKeyToken(Token)`).

For the getter (whether it's magic constant or not), we'll have to iterate over the variable names and pick the first one that matches `token(...)`, we can't do much better because we don't have access to table schema at this point.

I think it's worth leaving the "i => token(i)" logic in `getToken(String)`.

olim7t replied:
> getPartitionKeyToken() and setPartitionKeyToken(Token)

Or even simpler, `getToken()` and `setToken(Token)`, which have the advantage of being closer to the regular methods when the user does IDE completion.

pcmanus replied:
> As an alternative to the magic constant, we could have dedicated methods

Yes, that's probably even cleaner.

> Or even simpler, `getToken()` and `setToken(Token)`

I think I prefer `getPartitionKeyToken()/setPartitionKeyToken()` as it's more explicit what they do get/set. I'll admit that it's likely most people won't get confused if we use the order name, because again there isn't much use to get/set the token of anything else than the parititon key, but the more explicit naming could avoid confusing a couple users, and to some extend it's a better documentation of what the method does. I don't feel terribly strongly though.

> I think it's worth leaving the "i => token(i)" logic in `getToken(String)`.

Again, I worry that the inconsistency with {{setToken}} will surprise people, and I'd prefer limiting the magic to as many places as possible. If we do add `getPartitionKeyToken()/setPartitionKeyToken()`, then the only valid reason to even have `getToken(String)/setToken(String, Token)` is for when the user uses an alias (`getToken`) or a named bind marker (`setToken`). So I'd be in favor of reverting `getToken(String)/setToken(String, Token)` to their initial behavior (which is simple and consistent) and be very explicit in their javadoc that you should favor `getPartitionKeyToken()/setPartitionKeyToken()` unless you're using an alias/named bind marker.

olim7t replied:
Pushed to 2.0 and 2.1 as a separate commit (I had already merged the pull request).

tolbertam committed:
Add more Integration Tests for JAVA-312.

Added Integration Tests for clusters using VNodes.

Added Integration Test for verifying Metadata is updated when nodes are
added and removed.

Added CCMBridge#remove(n) for removing a node from the cluster.

Added more tests to TokenIntegrationTest.

Added more unit tests to TokenRangeTest.

olim7t commented:
Force-pushed to clean up history.
